### ISSUE TITLE
Allowing graphml as input

### ISSUE THREAD

**Issue by idshklein** [luukvdmeer/sfnetworks#72]
Given that I have a graphml file, it is not possible to load it as an sfnetworks object. 

The feature I'm requesting is creating a function which takes a graphml file as input, the fields of lat lon/xy of nodes, wkt geometry from edges and crs as parameters and output an sf_networks object. 

reproducible example:
```r
library(tidyverse)
library(igraph)
library(tidygraph)
library(sfnetworks)
library(ggraph)
library(ggspatial)
library(sf)
library(plotly)
G <- read_graph("Austria_highvoltage.graphml",format = "graphml")
G <- as_tbl_graph(G)
# nodes <- G %>% 
#   activate(nodes) %>%  
#   as_tibble() %>% 
#   st_as_sf(wkt = "wktsrid4326", crs = 4326)
edges <- G %>% 
  activate(edges) %>%  
  as_tibble() %>% 
  st_as_sf(wkt = "wktsrid4326", crs = 4326)
# austria_power <- bind_rows(nodes,edges)
net <- as_sfnetwork(edges)
```
**Describe alternatives you've considered**
A clear and concise description of any alternative solutions you've considered, possibly accompanied by a reproducible example with the [reprex package](https://reprex.tidyverse.org/).

**Comment by loreabad6** [comment]
Hi @idshklein, thank you for your request. To be able to test your example we would need the data. I found a file with the same name here: https://zenodo.org/record/250427#.X1Cbt3n7RaQ, can you confirm this is the data you are using? 

As I see from your reprex, this data can be transformed into a `tbl_graph` which then makes it possible to use with `sfnetworks`, given it has a geometry column. In your same example, the way to build the network would be as follows: 

``` r
library(tidyverse)
library(igraph)
library(tidygraph)
library(sfnetworks)
library(ggraph)
library(ggspatial)
library(sf)
library(plotly)

G <- read_graph("Austria_highvoltage.graphml",format = "graphml")
G <- as_tbl_graph(G)
nodes <- G %>%
  activate(nodes) %>%
  as_tibble() %>%
  st_as_sf(wkt = "wktsrid4326", crs = 4326)
edges <- G %>% 
  activate(edges) %>%  
  as_tibble() %>% 
  st_as_sf(wkt = "wktsrid4326", crs = 4326)
# austria_power <- bind_rows(nodes,edges)
net <- sfnetwork(nodes, edges)
#> Checking validity of network structure... Use force=TRUE to force construction without checks
#> Error in check_network_validity(nodes, edges, directed, edges_as_lines): Boundary points of edges should match their corresponding nodes
```

Now you see there is an error when creating the network, but this has to do with the data itself (if you take a closer look at it), since the coordinates on the geometry column for the nodes don't correspond to the order of the geometry column for the edges. 

A way around it might be to call the construction function with spatially implicit lines, and then convert the edges into spatially explicit with a morpher. 

``` r
(net <- sfnetwork(nodes,edges,edges_as_lines = F))
#> Checking validity of network structure... Use force=TRUE to force construction without checks
#> # An sfnetwork with 149 nodes and 169 edges
#> #
#> # CRS:  EPSG:4326 
#> #
#> # A directed acyclic simple graph with 2 components with spatially implicit edges
#> #
#> # Node Data:     149 x 14 (active)
#> # Geometry type: POINT
#> # Dimension:     XY
#> # Bounding box:  xmin: 9.591942 ymin: 46.53552 xmax: 16.99493 ymax: 48.5193
#>   netcapacity typ                 wktsrid4326 lat   voltage frequency ngen 
#>   <chr>       <chr>               <POINT [°]> <chr> <chr>   <chr>     <chr>
#> 1 ""          joint        (14.0444 48.08284) 48.0~ "22000~ ""        ""   
#> 2 ""          subs~         (12.5834 47.2785) 47.2~ ""      ""        ""   
#> 3 ""          subs~       (11.83871 47.39244) 47.3~ "22000~ "50"      ""   
#> 4 ""          joint       (12.13322 47.25083) 47.2~ "38000~ ""        ""   
#> 5 ""          joint       (16.44837 48.35864) 48.3~ ""      ""        ""   
#> 6 ""          subs~       (13.67984 46.54319) 46.5~ "22000~ ""        ""   
#> # ... with 143 more rows, and 7 more variables: operator <chr>, lon <chr>,
#> #   ref <chr>, source <chr>, name <chr>, capacity <chr>, id <chr>
#> #
#> # Edge Data: 169 x 17
#>    from    to rohmkm cables operator type  lengthm voltage frequency cnfkm
#>   <int> <int> <chr>  <chr>  <chr>    <chr> <chr>   <chr>   <chr>     <chr>
#> 1     1   122 ""     6      Austria~ line  444.42~ 220000  ""        ""   
#> 2     1    43 ""     6      Austria~ line  10974.~ 220000  ""        ""   
#> 3     2    31 ""     6      Austria~ line  9995.0~ 380000  ""        ""   
#> # ... with 166 more rows, and 7 more variables: fromrelation <chr>, ref <chr>,
#> #   name <chr>, xohmkm <chr>, wires <chr>, ithmaxa <chr>, lid <chr>
net %>% 
  convert(to_spatial_explicit_edges)
#> Error in UseMethod("st_as_sf"): nicht anwendbare Methode für 'st_as_sf' auf Objekt der Klasse "c('tbl_graph', 'igraph')" angewendet
```


However the last step seems to be having some problems, which I will need to look at later. Hope the solution so far helps, and I will come back to the issue if there is a need to report a bug soon.

**Comment by idshklein** [comment]
I confirm this is the same data

**Comment by luukvdmeer** [comment]
Any `igraph` object that has nodes which can be directly turned into an `sf` object with `st_as_sf` can be read by `as_sfnetworks`. The `...` parameters will be passed on to `st_as_sf` internally. That is, for graphml files:

```r
G = as_sfnetwork(
  x = igraph::read_graph("Austria_highvoltage.graphml", format = "graphml"),
  wkt = "wktsrid4326",
  crs = 4326
)
```
However, this creates only a network with the nodes being internally an sf object, not the edges. I agree that is something that we might need to consider changing and/or make it clearer. The way to go now is to explicitly add a geometry column to the edges after the initial network creation.

```r
G = activate(G, "edges")
st_geometry(G) = sf::st_as_sfc(tidygraph::pull(G, wktsrid4326))
```
In this specific case this will throw the same error as @loreabad6 found (i.e. 'Boundary points of edges should match their corresponding nodes'). That is a problem with the input data. You can work around that by forcing construction, and skipping this network validity checks. For most functions it will not be a problem that the coordinates don't match.

```r
nodes = G %>%
  activate(nodes) %>%
  st_as_sf()

edges = G %>% 
  activate(edges) %>%  
  as_tibble() %>% 
  st_as_sf(wkt = "wktsrid4326", crs = 4326)

net = sfnetwork(nodes, edges, force = TRUE)
```

So this is how it works now, but I seems to me the workflow can be improved.

----

Regarding this: 

> However the last step seems to be having some problems, which I will need to look at later.

This was indeed a bug and is fixed in `develop`. Will be pushed to master at next version release.

**Comment by luukvdmeer** [comment]
We will keep to rely on `igraph` for importing networks from graph specific file types like GraphML. However, a (hopefully) clear example of doing this is now added to the main vignette, see [here](https://luukvdmeer.github.io/sfnetworks/articles/structure.html#from-a-network-specific-file-type).

TL;DR: Use this:

```r
library(sfnetworks)
library(sf)
library(tidygraph)

url = 'https://raw.githubusercontent.com/ComplexNetTSP/Power_grids/v1.0.0/Countries/Netherlands/graphml/Netherlands_highvoltage.graphml'

igraph::read_graph(url, format = "graphml") %>%
  as_sfnetwork(wkt = "wktsrid4326", crs = 4326) %>%
  convert(to_spatial_explicit, wkt = "wktsrid4326", crs = 4326, .clean = TRUE)
```

It does require a clean spatial network structure (which for The Netherlands is true but for Austria not)
